### DISCUSSION TITLE
[Q/A][Java SDK (Component)] dubbo3.2.0 request header problem

### DISCUSSION THREAD

**Discussion by HollyWtk** [apache/dubbo#14279] [Question] [answered]
### Pre-check

- [X] I am sure that all the content I provide is in English.

### Apache Dubbo Component

Java SDK (apache/dubbo)

### Details

I want to combine SpringCloud Gateway and Dubbo REST protocol to do custom routing, and found that NettyHandler must capitalize the request header to get the request header, which conflicts with the request of Gateway, this is a bug or you need to handle the request header separately

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

**Comment by AlbumenJ** [comment]
Which version do you use?

**Reply by HollyWtk** [reply]
3.2.0  NettyHandler->dohandler()->acceptSupportJudge()->getAcceptMediaType()->RestHeaderEnum.ACCEPT

**Comment by suncairong163** [comment]
Do you mean SpringCloud Gateway request is not contain Accept header?  and the next route is Dubbo rest protocol request must have Accept header?

**Reply by suncairong163** [reply]
Accept header is null ,we will adjust Accept  to All (*/*)

**Reply by HollyWtk** [reply]
SpringCloud Gateway will change the request header to lowercase, while the enumeration of Dubbo source code is uppercase, e.g. accept=*/* will get the null value

**Reply by suncairong163** [reply]
> SpringCloud Gateway will change the request header to lowercase, while the enumeration of Dubbo source code is uppercase, e.g. accept=_/_ will get the null value

I have test this case ,it is support to get by lower case accept header ,and the soure support  code is


org.apache.dubbo.rpc.protocol.rest.request.NettyRequestFacade#getHeader


    public String getHeader(String name) {

        List<String> values = headers.get(name);

        if (values == null && name != null) {
            values = headers.get(name.toLowerCase());
        }

        if (values == null || values.isEmpty()) {
            return null;
        } else {
            return values.get(0);
        }
    }

**Reply by suncairong163** [reply]
and I can not recurrent your problem,  can you give  me  test case for me to recurrent?

**Reply by HollyWtk** [reply]
org.apache.dubbo.rpc.protocol.rest.handler#getAcceptMediaType
in dubbo-3.2.0
org.apache.dubbo.rpc.protocol.rest.request.NettyRequestFacade#getHeader
public String getHeader(String name) {

        List<String> values = headers.get(name);

        if (values == null || values.isEmpty()) {
            return null;
        } else {
            return values.get(0);
        }

    }
Is it inconsistent that the version we are talking about, I just checked the code for dubbo3.2.0 and the getHeader method does not become lowercase

**Reply by aofall** [reply, marked as answer]
> Is it inconsistent that the version we are talking about, I just checked the code for dubbo3.2.0 and the getHeader method does not become lowercase

This part of the code is included in version 3.2.8, you can try to upgrade the dubbo version. The latest version is 3.2.13

**Reply by HollyWtk** [reply]
ok thank you

**Comment by suncairong163** [comment]
ok，i will fix the bug

---- Replied Message ----
| From | ***@***.***> |
| Date | 06/07/2024 17:19 |
| To | apache/dubbo ***@***.***> |
| Cc | suncairong163 ***@***.***>,
Comment ***@***.***> |
| Subject | Re: [apache/dubbo] [Q/A][Java SDK (Component)] dubbo3.2.0 request header problem (Discussion #14279) |

SpringCloud Gateway will change the request header to lowercase, while the enumeration of Dubbo source code is uppercase, e.g. accept=/ will get the null value
